### PR TITLE
PWN-6090 - token selection connected to NewSend

### PR DESCRIPTION
**Pull request by ghost**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6090

## Description of Work

Token selection connected to NewSend

## Screenshot

Android Emulator - Resizable_API_33:5554 2022-12-07 14-36-28

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt [hunk 1]
```diff
@@ -8,15 +8,20 @@ import org.p2p.wallet.R
...
 import org.p2p.wallet.utils.Constants
+import org.p2p.wallet.utils.addFragment
 import org.p2p.wallet.utils.args
 import org.p2p.wallet.utils.viewbinding.viewBinding
 import org.p2p.wallet.utils.withArgs
 
 private const val ARG_RECIPIENT_ADDRESS = "ARG_RECIPIENT_ADDRESS"
 private const val ARG_RECIPIENT_USERNAME = "ARG_RECIPIENT_USERNAME"
 
+private const val EXTRA_TOKEN = "EXTRA_TOKEN"
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt:22]: KEY_RESULT_SOURCE_TOKEN

File: app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt [hunk 2]
```diff
@@ -46,18 +51,45 @@ class NewSendFragment :
...
     }
 
-    // TODO PWN-6090 connect this method to token selection
-    fun showSourceToken(token: Token.Active) {
+    private fun handleSupportFragmentResult(result: Bundle) {
+        when {
+            // will be more!
+            result.containsKey(EXTRA_TOKEN) -> {
+                val token = result.getParcelable<Token.Active>(EXTRA_TOKEN)
+                if (token != null) presenter.setSourceToken(token)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt:72]: ?.let?

File: app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt [hunk 3]
```diff
@@ -46,18 +51,45 @@ class NewSendFragment :
...
+            }
+        }
+    }
+
+    override fun showSourceToken(token: Token.Active) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt:77]: i don't think Source is a good naming in context of Send
in Swap, yes, we have source token and destination token which we want to swap, but in Send we only have token to send.
what do you think about showTokenToSend?

File: app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt [hunk 4]
```diff
@@ -46,18 +51,45 @@ class NewSendFragment :
...
             glideManager.load(imageViewTokenIcon, token.iconUrl)
             textViewTokenName.text = token.tokenSymbol
             textViewTokenTotal.text = token.getFormattedTotal(includeSymbol = true)
             textViewTokenAmountInUsd.text = token.getFormattedUsdTotal()
         }
     }
+
+    override fun navigateToTokenSelection(tokens: List<Token.Active>, selectedToken: Token.Active?) {
+        addFragment(
+            target = NewSelectTokenFragment.create(tokens, selectedToken, KEY_REQUEST_SEND, EXTRA_TOKEN),
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt:88]: named args

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt [hunk 5]
```diff
@@ -1,5 +1,45 @@
...
+    private val analyticsInteractor: ScreensAnalyticsInteractor,
+) : BasePresenter<NewSendContract.View>(), NewSendContract.Presenter {
+
+    private var token: Token.Active? by Delegates.observable(null) { _, _, newValue ->
+        if (newValue != null) view?.showSourceToken(newValue)
+    }
+
+    init {
+        launch {
+            token = userInteractor.getUserTokens().first()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:23]: we can see this screen only if user has tokens?

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt [hunk 6]
```diff
@@ -1,5 +1,45 @@
...
+import kotlin.properties.Delegates
 
-class NewSendPresenter() : BasePresenter<NewSendContract.View>(), NewSendContract.Presenter
+class NewSendPresenter(
+    private val userInteractor: UserInteractor,
+    private val browseAnalytics: BrowseAnalytics,
+    private val analyticsInteractor: ScreensAnalyticsInteractor,
+) : BasePresenter<NewSendContract.View>(), NewSendContract.Presenter {
+
+    private var token: Token.Active? by Delegates.observable(null) { _, _, newValue ->
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:17]: tokenToSend

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt [hunk 7]
```diff
@@ -1,5 +1,45 @@
...
+        if (newValue != null) view?.showSourceToken(newValue)
+    }
+
+    init {
+        launch {
+            token = userInteractor.getUserTokens().first()
+        }
+    }
+
+    override fun onTokenClicked() = loadTokensForSelection()
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:27]: block body if return type is unit

**Review by gslevinkov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt (hunk 5 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:23]: we just preselect SOL for this screen

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt (hunk 6 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:17]: no it is selected token in common

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:17]: but we don't send it? what this token for?

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt (hunk 7 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:27]: onClick() {
return Unit
}

Uhahaha

File: app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt (hunk 3 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendFragment.kt:77]: ok maybe

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt (hunk 6 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:17]: override fun setTokenToSend(newToken: Token.Active) {
        token = newToken
    }

**Review by 1sakov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt (hunk 5 above)
Comment (1sakov) [reply app/src/main/java/org/p2p/wallet/newsend/NewSendPresenter.kt:23]: Are you sure it's SOL? 
I remember we talked about stable coins here. 
Please talk to task owner

**Review by 1sakov — CHANGES_REQUESTED** [review]
1. The address should be ellipsized in the middle, not at the end
2. You missed one more textView under input
3. Enter the amount shouldn't be a slider
4. The fees icon should have text
5. The view should be little bit above the center, not exactly in the center
6. Missed token symbol at the end of the input

**Review by 1sakov — APPROVED** [review]

**Pull request merged**